onnxruntime/core/util/math_cpu.cc

```diff
@@ -80,6 +80,90 @@ void Gemm<float, ThreadPool>(CBLAS_TRANSPOSE TransA, CBLAS_TRANSPOSE TransB, ptr
   MlasGemm(TransA, TransB, M, N, K, alpha, A, lda, B, ldb, beta, C, N, threadpool);
 }
 
+template <>
+void Gemm<Eigen::half, ThreadPool>(CBLAS_TRANSPOSE TransA, CBLAS_TRANSPOSE TransB, ptrdiff_t M,
+                                   ptrdiff_t N, ptrdiff_t K, Eigen::half alpha, const Eigen::half* A, const Eigen::half* B, Eigen::half beta,
+                                   Eigen::half* C, ThreadPool*) {
+  auto C_mat = EigenMatrixMap<Eigen::half>(C, N, M);
+  if (beta == static_cast<Eigen::half>(0)) {
+    C_mat.setZero();
+  } else {
+    C_mat *= beta;
+  }
+  switch (TransA) {
+    case CblasNoTrans: {
+      switch (TransB) {
+        case CblasNoTrans:
+          C_mat.noalias() += alpha * (ConstEigenMatrixMap<Eigen::half>(B, N, K) *
+                                      ConstEigenMatrixMap<Eigen::half>(A, K, M));
+          return;
+        case CblasTrans:
+          C_mat.noalias() += alpha * (ConstEigenMatrixMap<Eigen::half>(B, K, N).transpose() *
+                                      ConstEigenMatrixMap<Eigen::half>(A, K, M));
+          return;
+        default:
+          ORT_THROW("CblasNoTrans Unexpected CBLAS_TRANSPOSE for TransB of ", TransB);
+      }
+    }
+    case CblasTrans: {
+      switch (TransB) {
+        case CblasNoTrans:
+          C_mat.noalias() += alpha * (ConstEigenMatrixMap<Eigen::half>(B, N, K) *
+                                      ConstEigenMatrixMap<Eigen::half>(A, M, K).transpose());
+          return;
+        case CblasTrans:
+          C_mat.noalias() += alpha * (ConstEigenMatrixMap<Eigen::half>(B, K, N).transpose() *
+                                      ConstEigenMatrixMap<Eigen::half>(A, M, K).transpose());
+          return;
+        default:
+          ORT_THROW("CblasTrans Unexpected CBLAS_TRANSPOSE for TransB of ", TransB);
+      }
+    }
+    default:
+      ORT_THROW("Unexpected CBLAS_TRANSPOSE for TransA of ", TransA);
+  }
+}
+
+#ifdef MLAS_F16VEC_INTRINSICS_SUPPORTED
+static bool IsMatrixZero(MLFloat16* C, ptrdiff_t size) {
+  if (C == nullptr) {
+    return true;
+  }
+  for (ptrdiff_t i = 0; i != size; ++i) {
+    if (C[i].val != 0) return false;
+  }
+  return true;
+}
+#endif
+template <>
+void Gemm<MLFloat16, ThreadPool>(CBLAS_TRANSPOSE TransA, CBLAS_TRANSPOSE TransB, ptrdiff_t M,
+                                 ptrdiff_t N, ptrdiff_t K, MLFloat16 alpha, const MLFloat16* A, const MLFloat16* B, MLFloat16 beta,
+                                 MLFloat16* C, ThreadPool* tp) {
+#ifdef MLAS_F16VEC_INTRINSICS_SUPPORTED
+  if (TransA == CblasNoTrans && TransB == CblasNoTrans && IsMatrixZero(C, M * N) && alpha.ToFloat() == 1.0 && beta.ToFloat() == 1.0) {
```

Comment:
>IsMatrixZero(C, M * N) [](http://example.com/codeflow?start=58&length=22)

The check of IsMatrixZero is heavy. We can move the call of MlasHalfGemmBatch to the Gemm::Compute, just like how we handle Gemm float32 there. 
If the condition meet, we call mlas or call the eigen library.

Reply:
If needed, we can first transpose and scale A and B, and then call MLAS API. It would definitely be faster than eigen.  MLAS api does not work well when the bias is not {N}, only then we can call eigen, but this almost never happens.

If B is const weight, we can even transpose and scale B during prepack.

Reply:
In almost all cases, it would be faster to call mlas api directly from the operator rather than going through this.